tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java

```diff
@@ -17,279 +17,332 @@
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
-import java.util.Properties;
 import java.util.Set;
 import java.util.function.Function;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static org.apache.kafka.clients.consumer.GroupProtocol.CLASSIC;
+import static org.apache.kafka.clients.consumer.GroupProtocol.CONSUMER;
+import static org.apache.kafka.common.ConsumerGroupState.EMPTY;
+import static org.apache.kafka.common.ConsumerGroupState.STABLE;
+import static org.apache.kafka.coordinator.group.GroupCoordinatorConfig.NEW_GROUP_COORDINATOR_ENABLE_CONFIG;
+import static org.apache.kafka.coordinator.group.GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG;
+import static org.apache.kafka.coordinator.group.GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
-public class DeleteConsumerGroupsTest extends ConsumerGroupCommandTest {
-    @ParameterizedTest
-    @ValueSource(strings = {"zk", "kraft"})
-    public void testDeleteWithTopicOption(String quorum) {
-        createOffsetsTopic(listenerName(), new Properties());
-        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--delete", "--group", GROUP, "--topic"};
-        assertThrows(OptionException.class, () -> getConsumerGroupService(cgcArgs));
-    }
-
-    @ParameterizedTest
-    @ValueSource(strings = {"zk", "kraft"})
-    public void testDeleteCmdNonExistingGroup(String quorum) {
-        createOffsetsTopic(listenerName(), new Properties());
-        String missingGroup = "missing.group";
 
-        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--delete", "--group", missingGroup};
-        ConsumerGroupCommand.ConsumerGroupService service = getConsumerGroupService(cgcArgs);
-
-        String output = ToolsTestUtils.grabConsoleOutput(service::deleteGroups);
-        assertTrue(output.contains("Group '" + missingGroup + "' could not be deleted due to:") && output.contains(Errors.GROUP_ID_NOT_FOUND.message()),
-            "The expected error (" + Errors.GROUP_ID_NOT_FOUND + ") was not detected while deleting consumer group");
+@ExtendWith(value = ClusterTestExtensions.class)
+@ClusterTestDefaults(clusterType = Type.ALL, serverProperties = {
+        @ClusterConfigProperty(key = OFFSETS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+        @ClusterConfigProperty(key = OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+        @ClusterConfigProperty(key = NEW_GROUP_COORDINATOR_ENABLE_CONFIG, value = "true")
```

Comment:
Passing this always true means that this test will never be executed with the legacy coordinator (which supports the classic protocol, with zk, and kraft). Not sure if I'm missing something but I would expect we want to run the test for both coordinators.  

We do test both coordinators in all consumer integration tests for example (see [here](https://github.com/apache/kafka/blob/7c0a302c4da9d53a8fddc504a9fac8d8afecbec8/core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala#L120) the combinations we currently run for them), and [here](https://github.com/apache/kafka/blob/7c0a302c4da9d53a8fddc504a9fac8d8afecbec8/core/src/test/scala/kafka/utils/TestInfoUtils.scala#L58) is how the `NEW_GROUP_COORDINATOR_ENABLE_CONFIG` ends up being determined based on the combinations.

Reply:
### server side

If `NEW_GROUP_COORDINATOR_ENABLE_CONFIG` gets true, kraft broker will create `GroupCoordinator`

https://github.com/apache/kafka/blob/1e8415160f96eb579ceaa3f89b3362f1deeccf6b/core/src/main/scala/kafka/server/BrokerServer.scala#L556

It enables to handle the requests used by `AsyncConsumer`. Also, `NEW_GROUP_COORDINATOR_ENABLE_CONFIG` do nothing to zk broker.

### client side

`GROUP_PROTOCOL_CONFIG` is used to pick up the impl of `Consumer`

https://github.com/apache/kafka/blob/1e8415160f96eb579ceaa3f89b3362f1deeccf6b/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerDelegateCreator.java#L62

> Passing this always true means that this test will never be executed with the legacy coordinator

not really since we still create `LegacyConsumer` to test the legacy coordinator by setting `group.protocol=classic`


> I would expect we want to run the test for both coordinators.

this PR loops both legacy and consumer protocol for testing `LegacyConsumer` and `AsyncConsumer`. Also, the kraft cluster supports requests of both consumers. Hence, it seems to me that is what we expect.


Or is there any risk of enabling `NEW_GROUP_COORDINATOR_ENABLE_CONFIG` to test `LegacyConsumer`?

Reply:
Hey @chia7712 , agree that when `NEW_GROUP_COORDINATOR_ENABLE_CONFIG` is true, kraft broker will use the new coordinator, with zk out of the pic, all good there. But

> we still create LegacyConsumer to test the legacy coordinator by setting group.protocol=classic

`LegacyConsumer` does not make the legacy coordinator kick in unless `isNewGroupCoordinatorEnabled` is false, because the classic protocol is supported by both coordinators. So when using the `LegacyConsumer`, it implies classic protocol (only one it supports), but the decision of which coordinator will serve the classic protocol is taken based on the broker config to enable the new coordinator or not. Since we're always enabling the new coordinator for this test, I expect we're never using the legacy coordinator created [here](https://github.com/apache/kafka/blob/1e8415160f96eb579ceaa3f89b3362f1deeccf6b/core/src/main/scala/kafka/server/BrokerServer.scala#L602)

> is there any risk of enabling NEW_GROUP_COORDINATOR_ENABLE_CONFIG to test LegacyConsumer

Not at all, actually we want to test it, my point is just that we also should test old coordinator + LegacyConsumer, and the way to achieve that is running the LegacyConsumer + NEW_GROUP_COORDINATOR_ENABLE_CONFIG=false;

Makes sense?

Reply:
> Not at all, actually we want to test it, my point is just that we also should test old coordinator + LegacyConsumer, and the way to achieve that is running the LegacyConsumer + NEW_GROUP_COORDINATOR_ENABLE_CONFIG=false;
Makes sense?

You are right. I overlooked the `GroupCoordinatorAdapter` :(

Reply:
@frankvicky @FrankYang0529 Could you please address @lianetm comments? This class and `DeleteOffsetsConsumerGroupCommandIntegrationTest` need to test "LegacyConsumer + NEW_GROUP_COORDINATOR_ENABLE_CONFIG=false (`GroupCoordinatorAdapter`)"

We can address that by `ClusterTemplate`. For example:

```java
    private static void generator(ClusterGenerator clusterGenerator) {
        Map<String, String> serverProperties = new HashMap<>();
        serverProperties.put(GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG, "1");
        serverProperties.put(GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, "1");
        // case 0: zk
        clusterGenerator.accept(ClusterConfig.defaultBuilder().setType(Type.ZK).setServerProperties(serverProperties).build());
        // case 1: kraft + legacy coordinator
        clusterGenerator.accept(ClusterConfig.defaultBuilder().setType(Type.KRAFT).setServerProperties(serverProperties).build());
        // case 2: combined kraft + legacy coordinator
        clusterGenerator.accept(ClusterConfig.defaultBuilder().setType(Type.CO_KRAFT).setServerProperties(serverProperties).build());

        serverProperties.put(GroupCoordinatorConfig.NEW_GROUP_COORDINATOR_ENABLE_CONFIG, "true");
        // case 3: kraft + new group coordinator
        clusterGenerator.accept(ClusterConfig.defaultBuilder().setType(Type.KRAFT).setName("newGroupCoordinator").setServerProperties(serverProperties).build());
        // case 4: combined kraft + new group coordinator
        clusterGenerator.accept(ClusterConfig.defaultBuilder().setType(Type.CO_KRAFT).setName("newGroupCoordinator").setServerProperties(serverProperties).build());
    }
```